### PR TITLE
Adding reencryption command to DKG

### PR DESCRIPTION
**Pull request by jbsv**

### CODE REVIEW

File: dkg/pedersen/pedersen.go [hunk 1]
```diff
@@ -488,6 +491,230 @@ func (w worker) work(jobIndex int) error {
 	return nil
 }
 
+// EncryptSecret implements dkg.Actor.
+func (a *Actor) EncryptSecret(msg []byte) (U kyber.Point, Cs []kyber.Point) {
```
Comment (pierluca) [inline dkg/pedersen/pedersen.go:495]: @si-co this is the function I was talking to you about.
Do you see how we could merge `EncryptSecret` and `VerifiableEncrypt` ?

File: dkg/pedersen/pedersen.go [hunk 2]
```diff
@@ -488,6 +491,230 @@ func (w worker) work(jobIndex int) error {
...
+				hiei := suite.Point().Mul(suite.Scalar().Neg(r.Ei), gxi)
+				hiHat := suite.Point().Add(gfi, hiei)
+				hash := sha256.New()
+				r.Ui.V.MarshalTo(hash)
+				uiHat.MarshalTo(hash)
+				hiHat.MarshalTo(hash)
+				e := suite.Scalar().SetBytes(hash.Sum(nil))
+				if e.Equal(r.Ei) {
+
+			*/
```
Comment (pierluca) [inline dkg/pedersen/pedersen.go:637]: do you need help fixing this ? @si-co offered to help :)

File: dkg/pedersen/pedersen.go [hunk 3]
```diff
@@ -488,6 +491,230 @@ func (w worker) work(jobIndex int) error {
...
+// Helper functions
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// DecryptSecret implements dkg.Actor.
+func (a *Actor) DecryptSecret(Cs []kyber.Point, XhatEnc kyber.Point, Sk kyber.Scalar) (msg []byte, err error) {
```
Comment (pierluca) [inline dkg/pedersen/pedersen.go:673]: Conceptually, shouldn't this logic be somewhere else ? 
If I understand correctly, this should be called by a client, normally...

File: dkg/pedersen/pedersen.go [hunk 4]
```diff
@@ -488,6 +491,230 @@ func (w worker) work(jobIndex int) error {
...
+	replies []types.ReencryptReply // replies received
+	Uis     []*share.PubShare      // re-encrypted shares
+}
+
+// newOCS creates a new on-chain secret structure.
+func newOCS(pubk kyber.Point) *OCS {
+	return &OCS{
+		pubk: pubk,
+	}
+}
```
Comment (pierluca) [inline dkg/pedersen/pedersen.go:539]: should the OCS-related logic be moved to its own file, for readability ?

**Review by pierluca — COMMENTED** [review]
First batch of review :)

File: dkg/pedersen/dkg.go [hunk 5]
```diff
@@ -301,20 +310,22 @@ func (s *instance) deal(ctx context.Context, out mino.Sender) error {
 			return xerrors.Errorf("context done: %v", ctx.Err())
 		}
 	}
+	//}
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/dkg.go:313]: whoopsy

File: dkg/pedersen/pedersen.go [hunk 6]
```diff
@@ -488,6 +491,230 @@ func (w worker) work(jobIndex int) error {
...
+		Cs = append(Cs, suite.Point().Add(C, kp))
+		dela.Logger.Debug().Msgf("Cs:%v", C)
+
+		msg = msg[min(len(msg), kp.EmbedLen()):]
+	}
+
+	return
+}
+
+type OCS struct {
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/pedersen.go:522]: agreed, thx

File: dkg/pedersen/pedersen_test.go [hunk 7]
```diff
@@ -225,6 +231,76 @@ func TestPedersen_Scenario(t *testing.T) {
...
+
+	fakeAuthority := NewAuthority(addrs, pubkeys)
+	_, err := actors[0].Setup(fakeAuthority, nbNodes)
+	require.NoError(t, err)
+
+	// every node should be able to encrypt/reencrypt/decrypt
+	message := []byte("Hello world")
+	kp := key.NewKeyPair(suites.MustFind("Ed25519"))
+
+	for i := 0; i < nbNodes; i++ {
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/pedersen_test.go:288]: for fun, I guess

File: dkg/pedersen/pedersen_test.go [hunk 8]
```diff
@@ -225,6 +231,76 @@ func TestPedersen_Scenario(t *testing.T) {
...
+		U, Cs := actors[i].EncryptSecret(message)
+
+		Uis, err := actors[i].ReencryptSecret(U, kp.Public, threshold)
+		require.NoError(t, err)
+		require.NotNil(t, Uis)
+
+		XhatEnc, err := share.RecoverCommit(suites.MustFind("Ed25519"), Uis, threshold, nbNodes)
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/pedersen_test.go:295]: agree, that's the "given" from cothority for now. This PR is a first step to have the calypso functionality in dela. But the interface is further worked on in the coming PR.

File: dkg/dkg.go [hunk 9]
```diff
@@ -29,6 +30,11 @@ type Actor interface {
 
 	Reshare(co crypto.CollectiveAuthority, newThreshold int) error
 
-	VerifiableDecrypt(ciphertexts []types.Ciphertext) ([][]byte, error)
 	VerifiableEncrypt(message []byte, GBar kyber.Point) (ciphertext types.Ciphertext, remainder []byte, err error)
+	VerifiableDecrypt(ciphertexts []types.Ciphertext) ([][]byte, error)
+
+	EncryptSecret(message []byte) (U kyber.Point, Cs []kyber.Point)
+	ReencryptSecret(U kyber.Point, Pk kyber.Point, threshold int) (Uis []*share.PubShare, err error)
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/dkg.go:37]: This is the first version and allows to have the calypso functionality in DELA. It's gonna be tidy up in the next version.

File: dkg/dkg.go [hunk 10]
```diff
@@ -29,6 +30,11 @@ type Actor interface {
...
+
+	DecryptSecret(Cs []kyber.Point, XhatEnc kyber.Point, Sk kyber.Scalar) (message []byte, err error)
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/dkg.go]: done

File: dkg/pedersen/controller/action.go [hunk 11]
```diff
@@ -81,14 +81,14 @@ type listenAction struct {
 }
 
 func (a listenAction) Execute(ctx node.Context) error {
-	var dkg dkg.DKG
+	var d dkg.DKG
 
-	err := ctx.Injector.Resolve(&dkg)
+	err := ctx.Injector.Resolve(&d)
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/controller/action.go:86]: done

File: dkg/pedersen/controller/controller.go [hunk 12]
```diff
@@ -67,6 +67,20 @@ func (m minimal) SetCommands(builder node.Builder) {
...
+	//	cli.StringFlag{
+	//		Name:  "encrypted",
+	//		Usage: "the encrypted string, as <hex(K)>:<hex(C)>",
+	//	},
+	//	cli.StringFlag{
+	//		Name:  "publickey",
+	//		Usage: "the new public key to reencrypt the message, as <hex(pk)>",
+	//	},
+	//)
+	//sub.SetAction(builder.MakeAction(reencryptAction{}))
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/controller/controller.go]: done

File: dkg/pedersen/dkg.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/dkg.go]: done

File: dkg/pedersen/pedersen.go (hunk 4 above)
Comment (jbsv) [reply dkg/pedersen/pedersen.go]: done

File: dkg/pedersen/pedersen_test.go [hunk 13]
```diff
@@ -225,6 +231,76 @@ func TestPedersen_Scenario(t *testing.T) {
...
+	fakeAuthority := NewAuthority(addrs, pubkeys)
+	_, err := actors[0].Setup(fakeAuthority, nbNodes)
+	require.NoError(t, err)
+
+	// every node should be able to encrypt/reencrypt/decrypt
+	message := []byte("Hello world")
+	kp := key.NewKeyPair(suites.MustFind("Ed25519"))
+
+	for i := 0; i < nbNodes; i++ {
+		U, Cs := actors[i].EncryptSecret(message)
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/pedersen_test.go:287]: done

File: dkg/pedersen/pedersen_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/pedersen_test.go]: done

File: dkg/dkg.go [hunk 14]
```diff
@@ -25,16 +24,32 @@ type Actor interface {
...
+	// Reshare recreates the DKG with an updated list of participants.
 	Reshare(co crypto.CollectiveAuthority, newThreshold int) error
 
+	// VerifiableEncrypt encrypts the given message into kyber points
+	// using the DKG public key and proof of work algorithm.
 	VerifiableEncrypt(message []byte, GBar kyber.Point) (ciphertext types.Ciphertext, remainder []byte, err error)
+
+	// VerifiableDecrypt decrypts a pair of kyber points into the original message
+	// using the DKG internal private key and a proof of work algorithm.
 	VerifiableDecrypt(ciphertexts []types.Ciphertext) ([][]byte, error)
```
Comment (pierluca) [inline dkg/dkg.go]: Proof-of-Work ? Are you sure about this one ? 😅 
My understanding is that we're just creating a zero-knowledge proof.

File: dkg/dkg.go [hunk 15]
```diff
@@ -25,16 +24,32 @@ type Actor interface {
 	// setup has not been done.
 	GetPublicKey() (kyber.Point, error)
 
+	// Encrypt encrypts the given message into kyber points
```
Comment (pierluca) [inline dkg/dkg.go:27]: I can see (here and below) that we're returning kyber points.
What do they contain ? What's K and C? Which one's the encrypted message and which one is the other stuff (randomness, etc.) ? 😅

File: dkg/pedersen/pedersen.go (hunk 3 above)
Comment (pierluca) [reply dkg/pedersen/pedersen.go:523]: I think we misunderstood each other 😅 

I didn't mean the OCS (although it's probably a good thing that you have split it out).

I meant the secret decryption seems off here (as we discussed in person yesterday), and it probably makes more sense into its own library.

**Review by pierluca — COMMENTED** [review]
Meta-comment:
I'm wondering, does it make sense to add Calypso-specific stuff to our Pedersen's implementation, or should we do a Calypso-specific DKG (in its own repo) using this implementation and extending it ? 🤔 
I'm just a bit worried of overspecializing our Pedersen implementation while at the same time slowing down our development (because I'll tend to be slightly more annoying on the Dela repo, since it's the foundation for everything else)

File: dkg/pedersen/ocs.go [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/share"
+	"go.dedis.ch/kyber/v3/suites"
+	"golang.org/x/net/context"
+	"golang.org/x/xerrors"
+)
+
+type onChainSecret struct {
+	U    kyber.Point // U is the random part of the encrypted secret
+	pubk kyber.Point // The client's public key
```
[the remark this answers is not in the document]
Comment (jbsv) [reply dkg/pedersen/ocs.go:17]: the client's

File: dkg/dkg.go (hunk 14 above)
Comment (jbsv) [reply dkg/dkg.go]: reworded

File: dkg/dkg.go (hunk 15 above)
Comment (jbsv) [reply dkg/dkg.go:27]: added description of K and C

**Pull request merged**